core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -212,13 +208,13 @@ public void initialize(InputSplit split, TaskAttemptContext context) throws IOEx
       recordTransformer = new AvroToStructuredTransformer();
     }
 
-    public StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder) throws IOException,
-      InterruptedException {
-      return recordTransformer.transform(recordBuilder, delegate.getCurrentKey().datum(), schema, pathField);
-    }
-
-    public StructuredRecord getCurrentValueNull() throws IOException, InterruptedException {
-      return recordTransformer.transform(delegate.getCurrentKey().datum(), pathField, path);
+    public StructuredRecord.Builder startCurrentValue() throws IOException, InterruptedException {
+      if (schema != null) {
+        return recordTransformer.transform(delegate.getCurrentKey().datum(), schema, pathField);
+      }
+      Schema schema = recordTransformer.generateStructuredSchema(delegate.getCurrentKey().datum());
+      return recordTransformer.transform(delegate.getCurrentKey().datum(), addFieldToSchema(schema, pathField),
```

Comment:
this will break when path field is null, you should only add the field if path is not null.